### PR TITLE
Holes are not correctly generated(Silent corruption)

### PR DESCRIPTION
**Pull request by pcd1193182**
Issue #4996 is ultimately caused because of a problem in the freeing logic. At the time when the dbufs are bzero'd, zfs doesn't yet have enough information to properly handle writes later in the same TXG.  By changing the code to use the new dbuf_write_children_ready code path in more cases, we address this issue.  The only time free_children will bzero a dbuf now is if we are freeing the entire dnode.

### CODE REVIEW

**Comment by behlendorf** [comment]
@pcd1193182 thanks for opening this.  Just to be clear is this the latest version of the proposed upstream fix?  If so it would be great if you could address the [cstyle warnings](http://build.zfsonlinux.org/builders/Debian%208%20arm%20%28BUILD%29/builds/7201/steps/shell_4/logs/stdio), add @rincebrain's test cases from 0605fb4f4210ef7572ce7ce6f8004ab9590ea7d1, and force update this PR.  That'll trigger a new build/test run and we can make sure we're reviewing the latest patch.

0605fb4f4210ef7572ce7ce6f8004ab9590ea7d1

**Comment by pcd1193182** [comment]
I have updated the branch; the tests have been added, and cstyle has been addressed.

**Comment by behlendorf** [comment]
@pcd1193182 thanks.  It looks like maybe you forgot to add the new files to the updated commit.

**Comment by pcd1193182** [comment]
Yep.  Sometimes I hate git.  Updated again.

**Comment by behlendorf** [comment]
You can run `make cstyle` locally before pushing.

**Comment by loyou** [comment]
two points should be said:
1. the patch sometimes [fail](http://build.zfsonlinux.org/builders/Amazon%202015.09%20x86_64%20%28TEST%29/builds/2350)  in #5004's build,  I made restrictive patch f8cdb29,  no matter if it is accepted, this fail should not be ignored.
2. more work(like descriptions and so on) should be done on @rincebrain's hole birth test case, it is modified from rsend, I have adjusted in 0605fb4, you can also refer it.

**Comment by pcd1193182** [comment]
It looks like the ztest failure occured because an assertion tripped in BP_SET_LEVEL; we tried to set the level to -1.  That would mean that the level for the db_blkptr is 0, which doesn't make sense, because the db_level is > 0.  I'd be happy to work on debugging this issue, but I'm not familiar with linux debugging tools.  Is there a core file that can be analyzed or anything?

**Comment by loyou** [comment]
This fail only happened in that build, I cannot reproduce on local test with --enable-debug.

**Comment by behlendorf** [comment]
@pcd1193182 so unfortunately right now the buildbot doesn't preserve core dumps from ztest.  That would be a nice future enhancement but right now the assumption is these same issues can be reproduced locally.  The good news is that there's really nothing special about the buildbot test environment.  They're stock installs of several common Linux distributions and all the [testing scripts](https://raw.githubusercontent.com/zfsonlinux/zfs-buildbot/master/scripts/bb-test-ztest.sh) are available from GitHub.

As for this specific ztest failure I think I've seen this same failure a handful of times on the master branch.  So I'm not positive it was introduced by this patch.  But it would be great if you could dig in to it.  My suggestion would be to [build ZFS in-tree](https://github.com/zfsonlinux/zfs/wiki/Building-ZFS), kick off `zloop.sh`, and just let it run overnight and see if you can reproduce it.  Then you'll locally have a core dump you can use gdb to analyze and unstripped binaries with full debugging symbols.  There's lots of documentation out there for gdb.

In the meanwhile can you please rebase this branch against the master branch and force update the branch on GitHub.  That'll trigger a fresh round of testing including a fresh run of `zloop.sh`.  You can also add the following like to your git commit comment to request that each tester run `zloop.sh` for an hour instead of just 15 minutes.

```
TEST_ZTEST_TIMEOUT=3600
```

**Comment by behlendorf** [comment]
@bprotopopov @loyou could you please review this.

**Comment by behlendorf** [comment]
@pcd1193182 you may be in luck.  The `replacement_001_pos` test failed on two of the automated testers in `dnode_dirty_l1range()`.  You can find the backtrace in the [console logs](http://build.zfsonlinux.org/builders/Amazon%202015.09%20x86_64%20SIMD%20%28TEST%29/builds/1823/steps/shell_11/logs/console) for analysis.  This _might_ be the same issue ztest encountered before.

**Comment by pcd1193182** [comment]
@behlendorf actually, that function (dnode_dirty_l1range) was created in this patch.  That's definitely a bug in this code.  We just hit that internally today, so I won't push this until we've figured this one out.

**Comment by bprotopopov** [comment]
The code LGTM as to the implementation of the core idea as described in the pull request.

File: module/zfs/dnode.c [hunk 1]
```diff
@@ -1815,6 +1880,8 @@ dnode_free_range(dnode_t *dn, uint64_t off, uint64_t len, dmu_tx_t *tx)
 		if (last != first)
 			dnode_dirty_l1(dn, last, tx);
 
+		dnode_dirty_l1range(dn, first, last, tx);
+
```
Comment (bprotopopov) [inline module/zfs/dnode.c]: Question for @pcd1193182 : what does the loop on line 1886 do that dnode_dirty_l1() and dnode_dirty_l1range() invocations just above did not accomplish ? It seems that the look below still deals with blocks from the same range ?

**Comment by bprotopopov** [comment]
Looked more at dnode_free_range() and have a question for @pcd1193182 (posted inline).

File: module/zfs/dnode.c (hunk 1 above)
Comment (pcd1193182) [inline module/zfs/dnode.c]: dnode_dirty_l1range dirties all incore dbufs.  The loop below dirties all dbufs that have been written to disk.  If you have some dbufs that were just created, they wouldn't be dirtied by the loop below.

**Comment by behlendorf** [comment]
Superseded by #5675.  The new PR includes the updated version of the patch from https://github.com/openzfs/openzfs/pull/273 and the test cases proposed in this PR.

**Pull request closed without merging**